### PR TITLE
docs: Add clarifications

### PR DESCRIPTION
**Pull request by CodeYan01**
Adds links to release functions for "new/incremented reference". Extra info on signal handler.
Parameter description on `obs_property_list_add_*` functions, since their parameter `name` can be confusing and it is unclear which value is stored in obs_data_t.
Adds clarifications on calldata_ptr.

### Description

See commit message above.

### Motivation and Context

Unclear to non-C programmers or beginners which ones need freeing.
I also got confused by obs_property_list_add_*.

### How Has This Been Tested?

Built locally and checked the sphinx features I haven't used before, like `:c:type:`.

### Types of changes

- Documentation (a change to documentation pages)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: docs/sphinx/reference-sources.rst [hunk 1]
```diff
@@ -945,7 +946,8 @@ General Source Functions
 
 .. function:: signal_handler_t *obs_source_get_signal_handler(const obs_source_t *source)
 
-   :return: The source's signal handler
+   :return: The source's signal handler. Should not be manually freed,
+            as its lifecycle is managed by the program.
```
Comment (notr1ch) [inline docs/sphinx/reference-sources.rst:950]: "the program" should probably be "the library" or "libobs" here for extra clarity.

File: docs/sphinx/reference-sources.rst [hunk 2]
```diff
@@ -954,7 +956,8 @@ General Source Functions
 
 .. function:: proc_handler_t *obs_source_get_proc_handler(const obs_source_t *source)
 
-   :return: The procedure handler for a source
+   :return: The procedure handler for a source. Should not be manually freed,
+            as its lifecycle is managed by the program.
```
Comment (notr1ch) [inline docs/sphinx/reference-sources.rst:960]: Same here.

File: docs/sphinx/reference-sources.rst (hunk 1 above)
Comment (CodeYan01) [reply docs/sphinx/reference-sources.rst]: Please see new commit for the changes. If you approve, I'll squash it. I added the same clarification to similar signal handler and proc handlers.

Comment (notr1ch) [reply docs/sphinx/reference-sources.rst]: Looks good to me!

**Comment by CodeYan01** [comment]
Thank you! Rebased and squashed.

**Pull request merged**